client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelInitializer.java

```diff
@@ -0,0 +1,154 @@
+   private static boolean trace = log.isTraceEnabled();
+
+   private final Bootstrap bootstrap;
+   private final SocketAddress unresolvedAddress;
+   private final OperationsFactory operationsFactory;
+   private final Configuration configuration;
+   private ChannelPool channelPool;
+   private volatile boolean isFirstPing = true;
+
+   ChannelInitializer(Bootstrap bootstrap, SocketAddress unresolvedAddress, OperationsFactory operationsFactory, Configuration configuration) {
+      this.bootstrap = bootstrap;
+      this.unresolvedAddress = unresolvedAddress;
+      this.operationsFactory = operationsFactory;
+      this.configuration = configuration;
+   }
+
+   CompletableFuture<Channel> createChannel() {
+      ChannelFuture connect = bootstrap.clone().connect();
+      ActivationFuture activationFuture = new ActivationFuture();
+      connect.addListener(activationFuture);
+      return activationFuture;
+   }
+
+   @Override
+   protected void initChannel(Channel channel) throws Exception {
+      if (trace) {
+         log.tracef("Created channel %s", channel);
+      }
+      if (configuration.security().ssl().enabled()) {
+         initSsl(channel);
+      }
+
+      AuthenticationConfiguration authentication = configuration.security().authentication();
+      if (authentication.enabled()) {
+         initAuthentication(channel, authentication);
+      }
+
+      if (configuration.connectionPool().minEvictableIdleTime() > 0) {
+         channel.pipeline().addLast(
+               new IdleStateHandler(0, 0, configuration.connectionPool().minEvictableIdleTime(), TimeUnit.MILLISECONDS),
+               new IdleStateHandlerProvider(configuration.connectionPool().minIdle(), channelPool));
+      }
+      ChannelRecord channelRecord = new ChannelRecord(unresolvedAddress, channelPool);
+      channel.attr(ChannelRecord.KEY).set(channelRecord);
+      if (isFirstPing) {
+         isFirstPing = false;
+         channel.pipeline().addLast(new InitialPingHandler(operationsFactory.newPingOperation(false)));
+      } else {
+         channel.pipeline().addLast(ActivationHandler.INSTANCE);
+      }
+   }
+
+   private void initSsl(Channel channel) {
+      SslConfiguration ssl = configuration.security().ssl();
+      SSLContext jdkSslContext = ssl.sslContext();
+      if (jdkSslContext == null) {
+         //noinspection deprecation
+         jdkSslContext = SslContextFactory.getContext(
+               ssl.keyStoreFileName(),
+               ssl.keyStoreType(),
+               ssl.keyStorePassword(),
+               ssl.keyStoreCertificatePassword(),
+               ssl.keyAlias(),
+               ssl.trustStoreFileName(),
+               ssl.trustStoreType(),
+               ssl.trustStorePassword(),
+               ssl.protocol(),
+               configuration.classLoader());
+      }
+      SslContext nettySslContext = new JdkSslContext(jdkSslContext, true, ClientAuth.NONE);
```

Comment:
I think there was a method called `SslContext#newClientContext` in Netty. Could we use it here?

Reply:
Probably yes, I've checked `SslUtils` in server and put the bits here. Should we use `SslContextBuilder` when we tie this to JDK impl anyway? (the method you mention is deprecated)

Reply:
I'm not entirely sure if we should tie to JDK implementation here. Sometimes you just want to trust all certificates and just encrypt the traffic. In that case you have no truststore and you can use OpenSSL implementation. BTW, while implementing Netty-based REST, I noticed that using OpenSSL implementation on the client side gave me huge performance boost. It's really worth exploring.

The Javadoc [here](https://github.com/netty/netty/blob/ab9f0a0fda23e9c254cca9b18f5c19090b6c63ef/handler/src/main/java/io/netty/handler/ssl/SslContext.java#L82) suggests that you should use `SslContextBuilder`. 

BTW, if you don't want to explore it here - just log a JIRA and assign it to me. I have some experience with Netty+SSL.